### DISCUSSION TITLE
how can i override default airflow.cfg in docker?

### DISCUSSION THREAD

**Discussion by iRecursion** [apache/airflow#26679] [Q&A] [answered]
### Environment： airflow 2.4.0 python3.8 ubuntu 20.04 lts

```Dockerfile
FROM airflow:240p38
  
# COPY ./airflow_docker.cfg airflow.cfg

RUN pip3 install --no-cache-dir apache-airflow-providers-celery==3.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-elasticsearch==4.2.0
RUN pip3 install --no-cache-dir apache-airflow-providers-http==4.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-imap==3.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-mysql==3.2.0
RUN pip3 install --no-cache-dir apache-airflow-providers-oracle==3.3.0
RUN pip3 install --no-cache-dir apache-airflow-providers-redis==3.0.0

RUN pip3 install cx_Oracle
```
when i comment the line `# COPY ./airflow_docker.cfg airflow.cfg`
it works,but enter the container,configuration not modified，such as follows:

````ini
[core]
dags_folder = /opt/airflow/dags
hostname_callable = airflow.utils.net.getfqdn
default_timezone = utc
executor = SequentialExecutor
```

**Comment by iRecursion** [comment]
* when i uncomment the line `COPY ./airflow_docker.cfg airflow.cfg`

```shell
airflow-airflow-init-1  | /home/airflow/.local/lib/python3.8/site-packages/airflow/configuration.py:363: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.
airflow-airflow-init-1  |   warnings.warn(
airflow-airflow-init-1  | Unable to load the config, contains a configuration error.
airflow-airflow-init-1  | Traceback (most recent call last):
airflow-airflow-init-1  |   File "/usr/local/lib/python3.8/logging/config.py", line 563, in configure
airflow-airflow-init-1  |     handler = self.configure_handler(handlers[name])
airflow-airflow-init-1  |   File "/usr/local/lib/python3.8/logging/config.py", line 744, in configure_handler
airflow-airflow-init-1  |     result = factory(**kwargs)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/log/file_processor_handler.py", line 45, in __init__
airflow-airflow-init-1  |     self.filename_template, self.filename_jinja_template = parse_template_string(filename_template)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/helpers.py", line 168, in parse_template_string
airflow-airflow-init-1  |     return None, jinja2.Template(template_string)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/environment.py", line 1208, in __new__
airflow-airflow-init-1  |     return env.from_string(source, template_class=cls)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/environment.py", line 1105, in from_string
airflow-airflow-init-1  |     return cls.from_code(self, self.compile(source), gs, None)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/environment.py", line 768, in compile
airflow-airflow-init-1  |     self.handle_exception(source=source_hint)
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/environment.py", line 936, in handle_exception
airflow-airflow-init-1  |     raise rewrite_traceback_stack(source=source)
airflow-airflow-init-1  |   File "<unknown>", line 1, in template
airflow-airflow-init-1  | jinja2.exceptions.TemplateSyntaxError: expected token ':', got '}'
airflow-airflow-init-1  | 
airflow-airflow-init-1  | The above exception was the direct cause of the following exception:
airflow-airflow-init-1  | 
airflow-airflow-init-1  | Traceback (most recent call last):
airflow-airflow-init-1  |   File "/home/airflow/.local/bin/airflow", line 5, in <module>
airflow-airflow-init-1  |     from airflow.__main__ import main
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/__init__.py", line 46, in <module>
airflow-airflow-init-1  |     settings.initialize()
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/settings.py", line 564, in initialize
airflow-airflow-init-1  |     LOGGING_CLASS_PATH = configure_logging()
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/logging_config.py", line 74, in configure_logging
airflow-airflow-init-1  |     raise e
airflow-airflow-init-1  |   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/logging_config.py", line 69, in configure_logging
airflow-airflow-init-1  |     dictConfig(logging_config)
airflow-airflow-init-1  |   File "/usr/local/lib/python3.8/logging/config.py", line 808, in dictConfig
airflow-airflow-init-1  |     dictConfigClass(config).configure()
airflow-airflow-init-1  |   File "/usr/local/lib/python3.8/logging/config.py", line 570, in configure
airflow-airflow-init-1  |     raise ValueError('Unable to configure handler '
airflow-airflow-init-1  | ValueError: Unable to configure handler 'processor'
airflow-airflow-init-1  | 
airflow-airflow-init-1  | ERROR!!!: Too old Airflow version !
airflow-airflow-init-1  | The minimum Airflow version supported: 2.2.0. Only use this or higher!
````

* **ERROR!!!: Too old Airflow version !**

* so my question is how to override airflow.cfg in Dockerfile,
  alternative option: .env file or using airflow-common-env in docker-compose.yaml?
  if every entry(key value paire) config in .env or docker-compose.yaml，
  that would get complicated and the airflow.cfg would be pointless

the same as: 

```Dockerfile
FROM airflow:240p38

USER root

RUN apt-get update \
  && apt-get install -y --no-install-recommends vim \
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*

USER airflow

RUN mkdir -p /opt/airflow
COPY ./airflow_docker.cfg /opt/airflow/airflow.cfg
ENV AIRFLOW_HOME=/opt/airflow

COPY ./airflow_docker.cfg airflow.cfg

RUN pip3 install --no-cache-dir apache-airflow-providers-celery==3.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-elasticsearch==4.2.0
RUN pip3 install --no-cache-dir apache-airflow-providers-http==4.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-imap==3.0.0
RUN pip3 install --no-cache-dir apache-airflow-providers-mysql==3.2.0
RUN pip3 install --no-cache-dir apache-airflow-providers-oracle==3.3.0
RUN pip3 install --no-cache-dir apache-airflow-providers-redis==3.0.0

RUN pip3 install cx_Oracle
```

* is there any other better solution, must build from source? thank you

**Comment by potiuk** [comment, marked as answer]
You have something srsly broken. You need to verify if you have the right version and observe the output of your Docker build process when you build docker image. Maybe even you have not build the image properly - It's hard to help you because you have not explained exactly what you do, what kind of logs you have when you build the image, how you build it, what is the airflow:240p38 really (does it have the right airflow in the first place).

Also Docker/Docker Compose require you to be pretty fluent when you want to modify and customize them, I am afraid. It requires a good deal of building and pushing image to registry,  checking the output of the build, understanding what happens there so if you are not ready for that - either do some investment first and learn about it all (and come back here when you explain step-by-step what you do and which of the errors/logs you do not understand) or switch to using Airflow from venv.

**Reply by iRecursion** [reply]
> You have something srsly broken. You need to verify if you have the right version and observe the output of your Docker build process when you build docker image. Maybe even you have not build the image properly - It's hard to help you because you have not explained exactly what you do, what kind of logs you have when you build the image, how you build it, what is the airflow:240p38 really (does it have the right airflow in the first place).
> 
> Also Docker/Docker Compose require you to be pretty fluent when you want to modify and customize them, I am afraid. It requires a good deal of building and pushing image to registry, checking the output of the build, understanding what happens there so if you are not ready for that - either do some investment first and learn about it all (and come back here when you explain step-by-step what you do and which of the errors/logs you do not understand) or switch to using Airflow from venv.

thanks,my objective is to extending official image,then install provider
,and custom airflow.cfg configuration，no other things

the image `airflow:240p38` just as new tag from official image `docker pull apache/airflow:2.4.0-python3.8`,
throw exception `ERROR!!!: Too old Airflow version` when using image `apache/airflow:2.4.0-python3.8`

I just used the way you provided,via `.env`,it works
.env has higher priority than airflow.cfg?

```shell
docker compose up airflow-init --env-file .env
docker compose --env-file .env up airflow-init
```

this can be used as a temporary solution, is it possible to externalize airflow.cfg?,
as a mount point，such as mysql(my.cnf), redis(redis.conf)
https://hub.docker.com/_/mysql
https://hub.docker.com/_/redis

**Reply by potiuk** [reply]
yes. You have all indormation about config in the official docs https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html. Just follow it and make sure you get it right (especially where you place the config).

**Reply by potiuk** [reply]
BTW. ERROR!!!: Too old Airflow version is printed by the Docker-compose (see yoursefl) - if you had it, it means that one of your changes causes (mistakenly) downgrade of airflow - possibly one of your `pip install` - you need to track it and see which one did it.
